tasks/deploy.py

```diff
@@ -0,0 +1,445 @@
+                    return line.strip()
+                else:
+                    log(f"{funcname}(): upload '{line.strip()}' did NOT match")
+            return None
+    else:
+        log(f"{funcname}(): upload log '{uploaded_txt}' does not exist")
+        return None
+
+
+def determine_successful_jobs(job_dirs):
+    """Determine all successful jobs provided in job_dirs.
+
+    Args:
+        job_dirs (list): list of job directories
+
+    Returns:
+        successes (list): list of dictionaries representing successful jobs
+    """
+    funcname = sys._getframe().f_code.co_name
+
+    successes = []
+    for job_dir in job_dirs:
+        slurm_out = determine_slurm_out(job_dir)
+        eessi_tarballs = determine_eessi_tarballs(job_dir)
+        if check_build_status(slurm_out, eessi_tarballs):
+            log(f"{funcname}(): SUCCESSFUL build in '{job_dir}'")
+            successes.append({'job_dir': job_dir,
+                              'slurm_out': slurm_out,
+                              'eessi_tarballs': eessi_tarballs})
+        else:
+            log(f"{funcname}(): FAILED build in '{job_dir}'")
+
+    return successes
+
+
+def determine_tarballs_to_deploy(successes, upload_policy):
+    """Determines tarballs to deploy depending on upload policy
+
+    Args:
+        successes (list): list of dictionaries
+                          {job_dir, slurm_out, eessi_tarballs}
+        upload_policy (string): one of 'all', 'latest' or 'once'
+            'all': deploy all
+            'latest': deploy only the last for each build target
+            'once': deploy only latest if none for this build target has
+                    been deployed before
+    Returns:
+        to_be_deployed (dictionary): dictionary of dictionaries
+                                     {job_dir, timestamp}
+    """
+    funcname = sys._getframe().f_code.co_name
+
+    log(f"{funcname}(): num successful jobs {len(successes)}")
+
+    to_be_deployed = {}
+    for s in successes:
+        # all tarballs for successful job
+        tarballs = s["eessi_tarballs"]
+        log(f"{funcname}(): num tarballs {len(tarballs)}")
+
+        # full path to first tarball for successful job
+        # Note, only one tarball per job is expected.
+        tb0 = tarballs[0]
+        log(f"{funcname}(): path to 1st tarball: '{tb0}'")
+
+        # name of tarball file only
+        tb0_base = os.path.basename(tb0)
+        log(f"{funcname}(): tarball filename: '{tb0_base}'")
+
+        # tarball name format: eessi-VERSION-COMPONENT-OS-ARCH-TIMESTAMP.tar.gz
+        # remove "-TIMESTAMP.tar.gz"
+        # build_target format: eessi-VERSION-COMPONENT-OS-ARCH
+        build_target = "-".join(tb0_base.split("-")[:-1])
+        log(f"{funcname}(): tarball build target '{build_target}'")
+
+        # timestamp in the filename
+        timestamp = int(tb0_base.split("-")[-1][:-7])
+        log(f"{funcname}(): tarball timestamp {timestamp}")
```

Comment:
Agree. The upload script creates a metadata file already. Question is a bit what is specific to the bot and what not. An example metadata file contains the following

```json
{
  "uploader": {
    "username": "trz42",
    "ip": "3.250.220.9",
    "hostname": "mgmt.fair-mastodon.citc.local"
  },
  "payload": {
    "filename": "eessi-2022.11-software-linux-x86_64-intel-haswell-1669128386.tar.gz",
    "size": "440965",
    "ctime": "Tue Nov 22 14:46:27 UTC 2022",
    "sha256sum": "b9265eda6da7cd908e58672dd0d91dbd16affa032d089ab22a5240deede7ec64",
    "url": "https://minio.nessi.no:60900/nessi-2022.11/2022.11/software/linux/x86_64/intel/haswell/1669128386/eessi-2022.11-software-linux-x86_64-intel-haswell-1669128386.tar.gz"
  },
  "link2pr": {
    "repo": "trz42/software-layer",
    "pr": "48"
  }
}
```

* `uploader` just records who and where a package was built. Maybe `uploader` could be changed to `builder` and additional fields could include information about the job (`job_id`, `working_directory`, ...)
* `payload` is mostly information about the tarball itself. The information encoded in the filename could be represented by additional fields: `repository`, `version`, `component`, `os`, `cpu family`, `microarchitecture`, `timestamp` ... then the filename could be just kept as is or it could be any arbitrary string.
* `link2pr` could be renamed to `source` or `origin`. There could be another field for `commit`.

All of the above, except the `url`, is available during the build step. Maybe `url` should be moved to a different entry `download`.

Reply:
See #120